cmd/viewer,types/views,various: avoid allocations in pointer field getters whenever possible

nickkhyl opened the pull request:
In this PR, we add a generic `views.ValuePointer` type that can be used as a view for pointers to basic types and struct types that do not require deep cloning and do not have corresponding view types. Its `Get` and `GetOk` methods return stack-allocated shallow copies of the underlying value.

We then update the `cmd/viewer` codegen to produce getters that return either concrete views when available or `ValuePointer` views when not, for pointer fields in generated view types. This allows us to avoid unnecessary allocations compared to returning pointers to newly allocated shallow copies.

Fixes #14570

raggi commented:
this is a generally good change, though I'd recommend starting to look at the corp integration asap in case you haven't already, as it will be much larger

bradfitz reviewed (commented):
Yay, thanks!

No viewer tests needed updating? Could you add some to lock this behavior in?

nickkhyl commented:
We don't really have proper tests for the codegens. Historically, I think we've relied on a [bunch of type definitions](https://github.com/tailscale/tailscale/blob/main/cmd/viewer/tests/tests.go) with various field types and generated view types and clone methods for them. If the viewer generates the code, and the code looks correct and builds, that's considered "tests passed" 😞.

But yeah.. We should probably start adding some real tests.

nickkhyl commented:
> We kinda do that by adding it to git already, though? Otherwise it shows up as dirty in our git status.

yep. And the `go_generate` CI [job](https://github.com/tailscale/tailscale/actions/runs/12681018599/job/35343899362) does just that:
 https://github.com/tailscale/tailscale/blob/a51672cafd8b6c4e87915a55bda1491eb7cbee84/.github/workflows/test.yml#L480-L491

bradfitz commented:
So can you add some code to `cmd/viewer/tests` that exercises your changed code paths?

nickkhyl commented:
It's already there. See `StructWithPtrs` in `cmd/viewer/tests/tests.go` and the updated `StructWithPtrsView.Value()` getter in `cmd/viewer/tests/tests_view.go`. It precisely covers the case in #14570:

```go
func (v StructWithPtrsView) Value() *StructWithoutPtrs {
	if v.ж.Value == nil {
		return nil
	}
	x := *v.ж.Value
	return &x
}
```
->
```go
func (v StructWithPtrsView) Value() StructWithoutPtrsView { return v.ж.Value.View() }
```

However, I also decided to update a more common case where the field is a pointer to a basic type or a trivial struct, such as `netip.Addr`. To handle this, I added a generic `views.ValuePointer`, whose `Value`/`ValueOk` methods return a stack-allocated shallow copy of the value, and updated `cmd/viewer/tests.StructWithPtrs` accordingly (see the new `NoView *StructWithNoView` field as well as the updated getter for the `Int *int` field).

nickkhyl commented on tailcfg/tailcfg_view.go:
```diff
@@ -145,21 +145,11 @@ func (v NodeView) Created() time.Time                       { return v.ж.Create
+	return views.ValuePointerOf(v.ж.LastSeen)
 }
 
-func (v NodeView) Online() *bool {
-	if v.ж.Online == nil {
-		return nil
-	}
-	x := *v.ж.Online
-	return &x
-}
```
This is the more common case I was referring to.

nickkhyl commented on ipn/ipnlocal/peerapi.go:
```diff
@@ -233,11 +233,11 @@ func (h *peerAPIHandler) logf(format string, a ...any) {
 // node originating from the peer.
 func (h *peerAPIHandler) isAddressValid(addr netip.Addr) bool {
-	if v := h.peerNode.SelfNodeV4MasqAddrForThisPeer(); v != nil {
-		return *v == addr
+	if v, ok := h.peerNode.SelfNodeV4MasqAddrForThisPeer().ValueOk(); ok {
+		return v == addr
 	}
-	if v := h.peerNode.SelfNodeV6MasqAddrForThisPeer(); v != nil {
-		return *v == addr
+	if v, ok := h.peerNode.SelfNodeV6MasqAddrForThisPeer().ValueOk(); ok {
+		return v == addr
 	}
 	pfx := netip.PrefixFrom(addr, addr.BitLen())
 	return views.SliceContains(h.selfNode.Addresses(), pfx)
```
Also, although this is orthogonal to the changes in this PR, it doesn't seem correct that `(*peerAPIHandler).isAddressValid` wouldn't check `SelfNodeV6MasqAddrForThisPeer` and `h.selfNode.Addresses()` if `SelfNodeV4MasqAddrForThisPeer` is non-nil. Is it a bug or am I missing something here?

cc @bradfitz

bradfitz replied:
Yeah, reads like a bug to me at least. I haven't looked to see what uses that code.

bradfitz replied:
Can you file a bug or send a separate PR?

nickkhyl replied:
https://github.com/tailscale/tailscale/pull/14637
